crypto/sm2/sm2_pmeth.c

```diff
@@ -63,6 +70,15 @@ static int pkey_sm2_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
             return 0;
         }
     }
+    if (sctx->id != NULL) {
+        dctx->id = OPENSSL_malloc(sctx->id_len);
+        if (dctx->id == NULL) {
+            pkey_sm2_cleanup(dst);
+            return 0;
+        }
+        memcpy(dctx->id, sctx->id, sctx->id_len);
+        dctx->id_len = sctx->id_len;
```

Comment:
Probably you need to copy `id_len` unconditionally to pick up the `-1` initialisation